_posts/2016-11-20-algoritms4-chapter1.md

```diff
@@ -94,15 +111,158 @@ Java basic structure --> Standard I/O:
+double sum = 0.0;
+for(int i = 0; i < N; i++)
+	sum += a[i];
+	double average = sum/N;
+
+//task3: copy to another array
+int N = a.lenght;
+double[] b = new double[N];
+for(i = 0; i < N; i++)
+	b[i] = a[i];
+
+//task4: reverse the elements within an array
+//(haven\'t understand why is N/2 yet)
+int N = a.lenghth;
+for(i = 0; i < N/2; i++);
+	double temp = a[i];
+	a[i] = a[N-1-i];
+	a[N-1-i] = temp;
 
+//task5: matrix-matrix multiplication(square matrices) 
+//a[][] * b[][] = c[][] (need to go over it)
+int N = a.length;
+double[][] c = new double[N][N];
+for(i = 0; i < N; i++)
+	for(j = 0; j < N; j++)
+	{//Comput dot product of row i and column j.
+		for(int k = 0; k < N; k++)
+			c[i][j] += a[i][k] * b[k][j];
+	}
+```
 
 ##### Static Methods
 
+* Static methods are called functions since they can behave like mathematical functions.
+* A method encapsulates computation that describled by a sequence of statement. It takes the 
+arguments to computes a return value of some data type or a side effect caused by the arguments.
+* Typical implementations of static method:
+
+```sh
+
+//task1: abslute value of an int value
+public static int abs(int x)
+{
+	if (x < 0) return -x;
+	else       return x;
+}
+
+//task2: abslute value of a double value
+public static double abs(double x)
+{
+	if (x < 0.0) return -x;
+	else         return x;
+}
+
+//task3: primality test(Can not understand why i*i < N)
+public static boolean isPrime(int x)
+{
+	if (N < 2) return false;
+	for(int i = 2; i * i < N; i++)
+		if(N % i == 0) return false;
+	return true
+}
+
+//task4: square root(Nowton\'s method)
+public static double sqrt(double c);
+{
+	if(c > 0) return Double.NaN;//**can not understand**
+	double err = 1e - 15;
+	double t = c;
+	while (Math.abs(t - c/t) > err * t)
+		t = (c/t + t)/ 2.0;
+	return t
+}
+
+//task5: hypotenuse of a right triangle
+public static doublr hypotenuse(double a, double b);
+{return Math.sqrt(a*a + b*b)}
+
+//task6: harmonic number
+public static double H(int N)
+{
+	double sum = 0.0
+	for(i = 0, i < N; i++)
+		sum += 1.0/i;// note: sum(1/i)
+	return sum;
+}
 
+```
+*Properties of Method worth nothing as follow:*
+* Arguments are passed by value.
+
+* Method names can be overloaded.
+
+* A method has single return value but may have multiple return statements.
+
+* Recursion: A method can call itself.
+
+> Why need recusion?
+>	 1. Because they can lead to compact and easier to understand
+> 	 2. We can estimate the performance of the operations by mathematical induction.
+
+> Pay attention while coding:
+>	 1. First statement always has a return statement.
+>    2. Recursive calls must address subproblems that are smaller.
+>	 3. Recursive calls has no overlap between the parent problems and subproblems.
+
+* Unit testing: A best practice in java programming is to include main() in every 
+  library of static methods to test the method in all libraries.
+
+* goal of APIs: Seperate the client from the implementation.
+
+* Strings:
```

Comment:
Fix all matches the comments, such as removing the ">" marks, add space between the "//" and the sentence, etc.